src/datachain/lib/dc.py

```diff
@@ -1955,14 +1997,34 @@ def filter(self, *args) -> "Self":
             )
             ```
         """
-        return super().filter(*args)
+        return self.clone(query=self._query.filter(*args))
 
-    @detach
     def limit(self, n: int) -> "Self":
         """Return the first n rows of the chain."""
-        return super().limit(n)
+        return self.clone(query=self._query.limit(n))
 
-    @detach
     def offset(self, offset: int) -> "Self":
         """Return the results starting with the offset row."""
-        return super().offset(offset)
+        return self.clone(query=self._query.offset(offset))
+
+    def count(self) -> int:
```

Comment:
are we removing them from the DatasetQuery?

Reply:
No, these should stay in DatasetQuery for now, I think DataChain shouldn't know anything about SQL.